Refactor scripts to set up Rust on dev desktops

jdno opened the pull request:
The scripts to set up Rust for a user on the dev desktops have been refactored for a better onboarding experience and error handling. The init script installs rustup now, and can be extended in the future to also set the user's git username and email.

For users who want to set up their own Rust toolchain, the setup_rust script clones their fork of the rust-lang/rust repository, compiles it, and links the build artifacts. This is optional, however, and not required if users want to use the dev desktop to work with different repositories.

jdno commented:
The `init` and `setup_rust` scripts have been tested on the new instances and seem to work fine. :+1:

jdno commented on ansible/roles/dev-desktop/files/scripts/init.sh:
```diff
@@ -1,14 +1,15 @@
 
-setup_rustup.sh
+echo "Installing rustup..."
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
```
I am looking at https://github.com/rust-lang/simpleinfra/pull/110 as something that we can put into this script as well.

jdno commented on ansible/roles/dev-desktop/files/scripts/link_rust.sh:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+
+# Enable strict mode for Bash
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -euo pipefail
+IFS=$'\n\t'
```
Personally, I prefer repetition here since it makes the flags very explicitly. We can replace it with something like `source ./enable-strict-mode.sh`. But I'm afraid that no-one will know what that actually does, leading to confusion and frustration when debugging or extending a script.

jdno commented on ansible/roles/dev-desktop/files/scripts/link_rust.sh:
```diff
@@ -0,0 +1,27 @@
+  if [ -d "${D}" ]; then
+    pushd "${D}"
+
+    ./x.py build
```
I don't know enough about developing Rust to have an opinion on this. The previous version of the script failed since it couldn't find `stage1` or `stage2`, so I put the build command in there. With the `if` statement that follows the build command, we shouldn't get errors anymore even if Rust hasn't been built yet. So I'm fine removing the build command here.

jyn514 replied:
If we remove the build command, then this script won't do anything at all. My preference is the first thing I said, avoid x.py altogether and `touch` the relevant files so we can link a rustup toolchain without waiting for rustc to compiler.

jdno replied:
My assumption is that these scripts are not designed with onboarding in mind, but instead to help folks when they're already working on the dev desktop. While the script doesn't to anything on a fresh installation, it is probably useful when you've manually compiled stages 1 or 2 in your own fork?

jdno commented on ansible/roles/dev-desktop/files/scripts/link_rust.sh:
```diff
@@ -0,0 +1,27 @@
+      rustup toolchain link "$D"_stage1 "$D/build/$target/stage1"
+    fi
+
+    if [[ -d "$D/build/$target/stage2" ]]; then
+      rustup toolchain link "$D"_stage2 "$D/build/$target/stage2"
+    fi
```
The script gets called from at least one of the worktree scripts as well, so I'm assuming it's in there for cases where people have built `stage2` manually.

jdno commented on ansible/roles/dev-desktop/files/scripts/setup_rust.sh:
```diff
@@ -0,0 +1,29 @@
+git checkout upstream/master
+popd
+
+set_defaults.sh
+link_rust.sh
```
@jyn514 Since we removed `./x.py build` from `link_rust.sh`, should we put it in here? Without it, calling `link_rust.sh` is kinda useless.

jyn514 commented:
@jdno have you seen https://github.com/rust-lang/simpleinfra/pull/132#discussion_r1000658197 ?

jyn514 commented:
Looks great, thanks! My last comment is that https://github.com/rust-lang/rust/pull/103286 is landing soon - we may want to either wait until it lands so we can use the new sysroot names unconditionally, or use version detection so this doesn't silently break.

jyn514 commented:
@jdno try this:
```
; grep 'pub const VERSION' src/bootstrap/lib.rs | grep -o '[0-9]*'
2
```
I can add a comment to bootstrap saying to keep the version in the same file. Major version 3 is the one that's going to change the sysroot paths; there's not yet a way to test it unless you want to run your script against a checkout of https://github.com/rust-lang/rust/pull/103286.

Looking at the compiler version won't help because it just show `1.66-dev`; theoretically you could look at the git hash but that will only work once https://github.com/rust-lang/rust/pull/103286 has landed.

jdno commented:
I've added the `grep` command and an `if` block to handle the two different versions. Also rebased the branch to resolve a merge conflict after merging #110.
